src/prefect/runtime/flow_run.py

```diff
@@ -29,14 +30,36 @@ def __getattr__(name: str) -> Any:
 
         from prefect.runtime.flow_run import id
     """
-    env_key = f"PREFECT__RUNTIME__FLOW_RUN__{name.upper()}"
-    if env_key in os.environ:
-        return os.environ[env_key]
+
     func = FIELDS.get(name)
-    if func is None:
-        raise AttributeError(f"{__name__} has no attribute {name!r}")
+
+    # if `name` is an attribute but it is mocked through environment variable, the mocked type will be str,
+    # which might be different from original one. For consistency, cast env var to the same type
+    env_key = f"PREFECT__RUNTIME__FLOW_RUN__{name.upper()}"
+
+    if func is not None:
+        real_value = func()
+        if env_key in os.environ:
+            mocked_value = os.environ[env_key]
+            # cast `mocked_value` to the same type than `real_value`
+            if isinstance(real_value, bool):
+                return bool(mocked_value)
+            elif isinstance(real_value, int):
+                return int(mocked_value)
+            elif isinstance(real_value, float):
+                return float(mocked_value)
+            elif isinstance(real_value, pendulum.DateTime):
+                return pendulum.instance(dateparser.parse(mocked_value))
+            else:
+                # default str
+                return mocked_value
```

Comment:
Instead of every type being an if statement, we could create a type casting dictionary to make it more readable:

```python
type_cast = {
            bool: bool,
            int: int,
            ...
}
```

and then we can cast it in the end

```python
cast_func = type_cast.get(type(real_value), str)
        return cast_func(mocked_value)
```

Reply:
That's a great idea! I implemented it in that way, but without the default str. The reason is that an attribute is of type complex (i.e. a List, or Dict) then it is difficult to implement the casting from a string (we should define a language to do that, think about a dict of parameters, each value of a different type). With that in mind, I think it is better for the end user to write in the docs that complex types (lists and dicts for now) cannot be mocked using environment variables, and suggesting to do that using monkeypatch or a similar tool.

Reply:
Handling the built-in types for now makes sense 👍